io_u.c

```diff
@@ -680,7 +680,27 @@ static enum fio_ddir rate_ddir(struct thread_data *td, enum fio_ddir ddir)
 	if (td->o.io_submit_mode == IO_MODE_INLINE)
 		io_u_quiesce(td);
 
+	if (td->o.timeout && ((usec + now) > td->o.timeout))
+		usec = td->o.timeout - now;
+	
+	/*
+	 * Check if the usec is taking negative values.
+	 * However it is unlikely, But still a possibility.
+	 * If this is not taken care, usec_sleep() can take a 
+	 * huge nap which is capable of not exiting the process indefinitely
+	 * although the process/thread may be sleeping.
+	 */
+	if (usec == (1 << (sizeof(usec) - usec))) {
+		ddir = DDIR_INVAL;
+		return ddir;
+	}
```

Comment:
You only need to check for negativity in the case where we need to do adjustment (i.e. around 684). At that point you can simply check if now is bigger than td->o.timeout. If it is then you can return immediately with an invalid ddir as it's clear we don't want to sleep. If it's not then you go on and do the subtraction.

Reply:
I'm not sure if I'm doing anything different here although I agree that yours is more efficient.

I'm checking for negative value here. However, the fact that ```usec``` is of type ```unsigned int```, I'm performing 2's compliment to check if the value is negative.

However, if you feel that earlier comment suits to this context, I can fix that

Reply:
@sribs I agree the outcome is hopefully the same but in the case that timeout adjustment is not needed we will still evaluate the diff with your existing code. Also my case theoretically will handle very big values of usec where multiplying by two would cause overflow (but I think we've got other problems if usec is that big :-) ).

It's minor but I think
```c
if (now > td->o.timeout)
```
is easier to read than
```c
if (usec == (1 << (sizeof(usec) - usec))) {
```

and will need less comment.